src/Host/Client/Impl/Transports/WebSocketMessageTransport.cs

```diff
@@ -2,52 +2,62 @@
 // Licensed under the MIT License. See LICENSE in the project root for license information.
 
 using System;
+using System.IO;
 using System.Net.Sockets;
+using System.Net.WebSockets;
 using System.Threading;
 using System.Threading.Tasks;
 using System.Threading.Tasks.Dataflow;
-using WebSocketSharp;
-using WebSocketSharp.Server;
+using Microsoft.R.Host.Protocol;
 
 namespace Microsoft.R.Host.Client {
-    internal class WebSocketMessageTransport : WebSocketBehavior, IMessageTransport {
+    internal class WebSocketMessageTransport : IMessageTransport {
         private readonly WebSocket _socket;
         private readonly BufferBlock<Task<Message>> _incomingMessages = new BufferBlock<Task<Message>>();
         private readonly SemaphoreSlim _sendLock = new SemaphoreSlim(1, 1);
 
-        private WebSocket Socket => _socket ?? Context.WebSocket;
+        private WebSocket Socket => _socket;
 
-        public event EventHandler Open;
-        public event EventHandler Close;
-
-        public WebSocketMessageTransport() {
-            Protocol = "Microsoft.R.Host";
-        }
-
-        public WebSocketMessageTransport(WebSocket socket) : this() {
+        public WebSocketMessageTransport(WebSocket socket) {
             _socket = socket;
-            _socket.OnError += (sender, args) => OnError(args);
-            _socket.OnMessage += (sender, args) => OnMessage(args);
-            _socket.OnClose += (sender, args) => OnClose(args);
         }
 
         public async Task<Message> ReceiveAsync(CancellationToken ct = default(CancellationToken)) {
-            return await await _incomingMessages.ReceiveAsync(ct);
+            const int blockSize = 0x10000;
+            var buffer = new MemoryStream(blockSize);
+
+            while (true) {
+                int index = (int)buffer.Length;
+                buffer.SetLength(index + blockSize);
+
+                WebSocketReceiveResult wsrr;
+                try {
+                    wsrr = await _socket.ReceiveAsync(new ArraySegment<byte>(buffer.GetBuffer(), index, blockSize), ct);
+                } catch (IOException ex) {
+                    throw new MessageTransportException(ex);
+                } catch (SocketException ex) {
+                    throw new MessageTransportException(ex);
+                } catch (WebSocketException ex) {
+                    throw new MessageTransportException(ex);
+                }
+
+                buffer.SetLength(index + wsrr.Count);
+
+                if (wsrr.CloseStatus != null) {
+                    throw new OperationCanceledException("Connection closed by host.");
```

Comment:
That's the one that RHost throws to its clients, but it doesn't expect to see it from the transport. Should probably be MessageTransportException.

---
In reply to: [74971825](https://github.com/Microsoft/RTVS/pull/2173#discussion_r74971825) [](ancestors = 74971825)